rfcs/OptInFeatures.md

```diff
@@ -0,0 +1,152 @@
+  
+
+# RFC: Opt-in features
+
+ This document is a work in progress. A lot of the questions about introspection are closely related to [#300](https://github.com/graphql/graphql-spec/issues/300) (`Expose user-defined meta-information via introspection API in form of directives`) and will therefore need to be revisited based on the progress there.
+
+## 📜 Problem Statement
+
+  
+
+GraphQL has a [built-in mechanism for deprecation](https://spec.graphql.org/draft/#sec--deprecated) allowing to gracefully remove features from the schema. The lifecycle of a feature can typically be represented as `stable` -> `deprecated` -> `removed`.  
+
+In a lot of cases though, a feature lifecycle includes an experimental phase where it has just been added and can be changed without warning. In this state, the feature is usable and feedback is encouraged but isn't considered stable enough to be put in production. The feature lifecycle becomes `experimental` -> `stable` -> `deprecated` -> `removed` .
+
+  
+### Goals
+The goal of this proposal is to support the `experimental` state and, moving forward, any state that requires the client developer to make an explicit decision before using a given feature. In that sense, it's about "opting in" to using the feature, which includes supporting `experimental` states. 
+
+To give a few examples:
+
+* a field is experimental and might be changed or removed without prior notice (the above example).
+
+* a field is expensive to compute and should be used with caution.
+
+* a field has specific security requirements and requires a specific header or other form of authentication.
+
+### Non-goals
+This proposal is not about security and/or hiding parts of a schema. Its goal is to make it easier to communicate opt-in features to client developer and therefore needs to expose that information.
+
+## 👀 Prior work
+
+* GitHub uses [schema previews](https://docs.github.com/en/graphql/overview/schema-previews) to opt-in new features.
+* Kotlin has [OptIn requirements](https://kotlinlang.org/docs/opt-in-requirements.html) that started out as `@Experimental` before [being changed to `@OptIn`](https://youtrack.jetbrains.com/issue/KT-26216/Generalize-Experimental-API)
+
+## 🧑‍💻 Proposed solution
+
+### The `@optIn` directive
+
+It is proposed to add an `@optIn` directive to the specification:
+
+```graphql
+"""
+Indicates that the given field, argument, input field or enum value requires
+giving explicit consent before being used.
+"""
+directive @optIn(feature: String!) repeatable
```

Comment:
Should these be be `repeatable`? I'm not sure if that's necessary and it could complicate things

Reply:
Fair point. Since there are different features I figured out it was possible to require multiple of them and it didn't feel too complex but we can start simple and add `repeatable` down the road if we want to I guess. Although that circles back to the introspection discussion: if we add `repeatable`, then we need to model features as a list which might be breaking, etc...